crates/provider/src/chain/stream.rs

```diff
@@ -0,0 +1,159 @@
+    ) -> impl Stream<Item = Block> + 'static {
+        stream! {
+            let mut self_ = match Self::new(client, next_yield).await {
+                Ok(poller) => poller,
+                Err(err) => {
+                    warn!(%err, "Unable to start the stream");
+                    return;
+                }
+            };
+            loop {
+                let blocks = match self_.new_blocks().await {
+                    Ok(blocks) => blocks,
+                    Err(err) => {
+                        warn!(%err, "irrecoverable error, stopping the stream");
+                        break;
+                    }
+                };
+                for block in blocks {
+                    yield block;
+                    self_.next_yield += 1;
+                }
+            }
+        }
+    }
+
+    async fn new(
+        weak_client: WeakClient<T>,
+        next_yield: Option<BlockNumber>,
+    ) -> Result<Self, ChainStreamError<T>> {
+        let client = Self::client(&weak_client)?;
+        let chain_tip = client
+            .request("eth_blockNumber", ())
+            .await
+            .map(|b: U64| b.to::<u64>())
+            .map_err(|e| ChainStreamError::Rpc(e))?;
+        let next_yield = next_yield.unwrap_or(chain_tip);
+
+        let block_numbers_stream = PollerBuilder::new(weak_client.clone(), "eth_blockNumber", ())
+            .spawn()
+            .into_stream_raw();
+        Ok(Self {
+            client: weak_client,
+            next_yield,
+            chain_tip,
+            block_numbers_stream,
+            _phantom: PhantomData,
+        })
+    }
+
+    fn client(client: &WeakClient<T>) -> Result<Arc<RpcClientInner<T>>, ChainStreamError<T>> {
+        client.upgrade().ok_or(ChainStreamError::ClientDropped)
+    }
+
+    /// Updates the chain tip so that it is greater than `next_yield`.
+    async fn update_chain_tip(&mut self) -> Result<(), ChainStreamError<T>> {
+        // Loop until chain tip is advanced.
+        while self.next_yield > self.chain_tip {
+            let new_number = match self.block_numbers_stream.next().await {
+                Some(Ok(block_number)) => block_number.to::<u64>(),
+                Some(Err(err)) => {
+                    // This is fine.
+                    debug!(%err, "polling stream lagged");
+                    continue;
+                }
+                None => {
+                    debug!("polling stream ended");
+                    return Err(ChainStreamError::PollingStreamEnded);
+                }
+            };
+            self.chain_tip = new_number;
+        }
+        Ok(())
+    }
+
+    /// Tries to fetch the range of blocks that can surely be yielded via single batch request.
+    async fn fetch_blocks(&self) -> Result<Vec<Block>, ChainStreamError<T>> {
+        // We don't want to request too many batches by accident.
+        // TODO: Should be a property on transport or client.
+        const MAX_BATCH_SIZE: usize = 10;
```

Comment:
In most realistic cases, batches will be smaller than that, and if not, we remember `chain_tip` so we will send a new request on the next iteration without having to wait for the block number stream.

Still, probably it makes sense to make max batch size configurable, but it's out of scope for this PR.